### PR TITLE
[FIX] mrp: use cache in method explode

### PR DESCRIPTION
**Pull request by yelizariev**
this speeds up reading qty_available field in product templates with hundreds
variants.

Test for a product with 712 variants:

BEFORE: 1563 3.700 2.184
AFTER: 523 1.153 1.582

Description of the issue/feature this PR addresses:

Current behavior before PR:

Desired behavior after PR is merged:

--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

**Comment by yelizariev** [comment]
Related to https://github.com/odoo/odoo/pull/68637

**Comment by ryv-odoo** [comment]
Some comments:
- Don't use a cache in the context please. Because the context is immutable/hashable it will copy for each with_context. Why you don't use a facultative argument instead (it still ugly but better) ?
- Change in stable for a "small" improvements (I still a good one) seems not really blocking then I prefer to target the master for this (see next point).
- It is a little sad to add this complexity (code) only for compute `qty_xxx`. Maybe we can cache the `_bom_find` by product for each flow of mrp (not only for this compute). Crazy idea: create a `bom_phantom_id` (or/and `bom_id` for normal bom) compute field (contains the first matching bom) on `product.product` (only for master) with context_depends (like for qty_xxx). It allows to use the orm cache and the prefetch. And this compute could use the batch version of `_bom_find`. What do you think ?

**Comment by yelizariev** [comment]
> Some comments:
> 
> * Don't use a cache in the context please. Because the context is immutable/hashable it will copy for each with_context. Why you don't use a facultative argument instead (it still ugly but better) ?

Both are ugly, but adding an argument looks less robust for me. 

> * Change in stable for a "small" improvements (I still a good one) seems not really blocking then I prefer to target the master for this (see next point).

The customer experience 6 sec delay every time they open Products menu. It seems reasonable to update in stable if the changes are safe enough

> * It is a little sad to add this complexity (code) only for compute `qty_xxx`. Maybe we can cache the `_bom_find` by product for each flow of mrp (not only for this compute). Crazy idea: create a `bom_phantom_id` (or/and `bom_id` for normal bom) compute field (contains the first matching bom) on `product.product` (only for master) with context_depends (like for qty_xxx). It allows to use the orm cache and the prefetch. And this compute could use the batch version of `_bom_find`. What do you think ?

Looks interesting. Honestly I don't like this manual cache via context/new argument, and I'd prefer using field cache or a method with decorated ``ormcache``, but for stable is not always possible. 

I may try to update code in master once we decide what to do with this PR.

Btw, any plans to make populate methods for MRP/stock modules ?https://www.odoo.com/documentation/14.0/developer/reference/testing.html?highlight=populate#database-population

**Comment by ryv-odoo** [comment]
> Both are ugly, but adding an argument looks less robust for me.

Well, add an argument is really less robust ? In which case it can a issue (I don't make a lot of fix for the stable)?
Also it avoids to copy your cache for each with_context.

> The customer experience 6 sec delay every time they open Products menu. It seems reasonable to update in stable if the changes are safe enough

Even after the improvement on explode made few days before. If it is possible can you add a pyflam before/after ?

>Looks interesting. Honestly I don't like this manual cache via context/new argument, and I'd prefer using field cache or a method with decorated ormcache, but for stable is not always possible.
>I may try to update code in master once we decide what to do with this PR.

In master, a cleaner version is wanted.

> Btw, any plans to make populate methods for MRP/stock modules ?

Well, the populate exists already for stock/mrp/purchase/sale/product but only in master. Feel free to improve it (some part are messy, not complete).

**Comment by Aurelienvd** [comment]
> Even after the improvement on explode made few days before. If it is possible can you add a pyflam before/after ?

Yes. In a database with lots of variants (~1000) each one with a phantom BOM, the previous explode optimization gives a 2-3 speedup but it can still takes around 20s for the product.template search_read.

**Comment by Aurelienvd** [comment]
Question:

Since the cache is a mapping `{product -> bom}`, wouldn't there be an issue in the following scenario:
2 Companies 1 and 2
1 Product "Wood Plank" with `company_id` = False
1 Bom for "Wood Plank" with `company_id = 1`, 1 Bom for "Wood Plank" with `company_id = 2`
1 Product "Stool" with `company_id = 1`, 1 Product "Desk" with `company_id = 2`
1 Bom for "Stool" with `company_id = 1` and `bom_line_ids.product_id = Wood Plank`, 1 Bom for "Desk" with `company_id = 2` and `bom_line_ids.product_id = Wood Plank`

When computing the `virtual_available` for `self = product.product("Stool", "Desk", "Wood Plank")`,
exploding the "Stool" bom (`company_id = 1`) will set `product_boms = {"Stool": mrp.bom(3), "Wood Plank": mrp.bom(1)}`. Then, when exploding the "Desk" bom (`company_id = 2`), it will use the bom of "Wood Plank" in the cache, even though `mrp.bom(1).company_id = 1`.

In other words, in a multi-company case, with 1 product `company_id = False` and 1 bom for this product for each company in the database, couldn't the cache sometimes "bypass" the `company_id` filter in `_get_product2bom` ?

**Comment by yelizariev** [comment]
> Well, add an argument is really less robust?

If someone (including third-party developers) inherits that method without **kwargs, they will get an error *unknown argument*

```
 # their code
 def explode(self, product, quantity, picking_type=False):
      # ...
      return super().exlode(product, quantity, picking_type=product_type)

 # our code
 boms, bom_sub_lines = bom_kits[product].explode(product, 1, cache=cache)
```

Using **kwargs is not panacea either

```
 # their code
 def explode(self, product, quantity, picking_type=False, **kwargs):
      # ...
      return super().exlode(product, quantity, picking_type=product_type, **kwargs)

 # someone else code
 boms, bom_sub_lines = bom_kits[product].explode(product, 1, picking_type, cache)  # -> error explode takes from 2 to 3 positional arguments, but 4 were given.
```

**Comment by yelizariev** [comment]
> Even after the improvement on explode made few days before. If it is possible can you add a pyflam before/after ?
> 
> >

You mean before and after this PR? Check it here: https://www.odoo.com/web#action=4043&cids=1&id=2491624&menu_id=4720&model=project.task&view_type=form

File: addons/mrp/models/mrp_bom.py [hunk 1]
```diff
@@ -249,7 +249,11 @@ def check_cycle(v, visited, recStack, graph):
             return False
 
         product_ids = set()
-        product_boms = {}
+        cache = self.env.context.get("mrp_explode_cache", {})
+        product_boms = cache.get('product_boms')
+        if product_boms is None:
+            product_boms = {}
```
Comment (MiquelRForgeFlow) [inline addons/mrp/models/mrp_bom.py:255]:
```suggestion
        product_boms = cache.setdefault('product_boms', {}).get('product_boms')
```
less lines

Comment (yelizariev) [reply addons/mrp/models/mrp_bom.py]: Fixed this way
image

-- I don't want to change cache dict, because it would just mean to change new temporary dictionary when context doesn't have the key

**Comment by yelizariev** [comment]
https://github.com/odoo/odoo/pull/70963#issuecomment-843167609

@Aurelienvd fixed

**Comment by yelizariev** [comment]
r+ ?

**Comment by yelizariev** [comment]
@odoo/rd-logistics

**Comment by yelizariev** [comment]
my cache tricks never get approval fast...

**Comment by yelizariev** [comment]
source

**Comment by yelizariev** [comment]
Also, opw-2522267

**Comment by yelizariev** [comment]
Hello guys! 
6 seconds to open a menu is not good. Dropping it to 3 seconds would increases user satisfaction more than just twice time, I think.

waiting

**Comment by Yenthe666** [comment]
Guys? Let's get some performance stuff moving plox ❤️

**Comment by yelizariev** [comment]
Hello @odoo/rd-logistics

**Comment by Yenthe666** [comment]
Maybe @amoyaux can make us happy :)

**Comment by yelizariev** [comment]
Close it because of no activity :cry:
